metadata-ingestion/src/datahub/ingestion/source/state/redundant_run_skip_handler.py

```diff
@@ -104,46 +99,178 @@ def create_checkpoint(self) -> Optional[Checkpoint[BaseUsageCheckpointState]]:
+        last_checkpoint = self.state_provider.get_last_checkpoint(
+            self.job_id, BaseTimeWindowCheckpointState
+        )
         if (
-            last_successful_pipeline_run_end_time_millis is not None
-            and cur_start_time_millis <= last_successful_pipeline_run_end_time_millis
-        ):
-            warn_msg = (
-                f"Skippig this run, since the last run's bucket duration end: "
-                f"{get_datetime_from_ts_millis_in_utc(last_successful_pipeline_run_end_time_millis)}"
-                f" is later than the current start_time: {get_datetime_from_ts_millis_in_utc(cur_start_time_millis)}"
+            last_checkpoint is None or last_checkpoint.state is None
+        ) or self.should_skip_this_run(cur_start_time_millis, cur_end_time_millis):
+            return cur_start_time, cur_end_time
+
+        suggested_start_time_millis: int = cur_start_time_millis
+        suggested_end_time_millis: int = cur_end_time_millis
+
+        last_run_start_time, last_run_end_time = self.get_last_run_time_window(
+            last_checkpoint
+        )
+
+        # In case of usage, it is assumed that start_time_millis (cur_start_time_millis as well as last_run_start_time) always conincides
+        # with floored bucket start time. This should be taken care of outside this scope.
+
+        if cur_start_time_millis >= last_run_start_time:
+            if cur_start_time_millis > last_run_end_time:
+                if allow_expand:
+                    # scenario of time gap between past successful run window and current run window - maybe due to failed past run
+                    # Should we keep some configurable limits here to decide how much increase in time window is fine ?
+                    suggested_start_time_millis = last_run_end_time
+                    logger.info(
+                        f"{self.job_id} : Expanding time window. Changing start time to  {get_datetime_from_ts_millis_in_utc(last_run_end_time)}"
+                    )
+                else:
+                    logger.warn(
+                        f"{self.job_id} : Observed gap in last run end time({get_datetime_from_ts_millis_in_utc(last_run_end_time)}) and current run start time({cur_start_time})."
+                    )
+            elif cur_end_time_millis > last_run_end_time and allow_reduce:
+                # scenario of scheduled ingestions with default start, end times
+                suggested_start_time_millis = last_run_end_time
+                logger.info(
+                    f"{self.job_id} : Reducing time window. Changing start time to  {get_datetime_from_ts_millis_in_utc(last_run_end_time)}"
+                )
+        else:
+            # cur_start_time_millis < last_run_start_time
+            # This is most likely a manual backdated run which we should always run.
+            # Do we really need below optimisation/reducing time window for manual runs ?
```

Comment:
I'm not sure here. On the one hand, maybe you want to run everything, on the other, perhaps you forgot when we last successfully ran usage and want to ingest a wide range in the past to be sure. I think I'd rather keep this optimization but provide a config option that sets `allow_reduce` to `False`

Reply:
allow_reduce here is a stateful usage/lineage specific config. The same discussion here - more options within stateful config may cause more confusion than help. I'll keep this as is in favour of simplicity. Will update comments here.